pkg/daemon/ceph/client/filesystem.go

```diff
@@ -196,26 +197,120 @@ if Ceph version is Luminous (12.y.z) and num active mdses (max_mds) was lowered,
-Using slightly stale info, this could (rarely) result in momentary loss of filesystem availability: %v`, fsName, err)
+		logger.Warningf(
+			fmt.Sprintf("Failed to get filesystem %s info needed to deactivate running mdses. ", fsName) +
+				"using slightly stale info, this could (rarely) result in momentary loss of filesystem availability" +
+				fmt.Sprintf(": %v", err),
+		)
 		fs = fsAtStart // <-- Do the best we can to disable mdses that were active when we started
 		// Effects of stale info should be non-destructive, & unlikely the info is actually bad
 	}
-	if len(fs.MDSMap.In) > int(activeMDSCount) {
-		// Only try to deactivate mdses if there are more than desired.
-		for _, gid := range fs.MDSMap.In[activeMDSCount:] {
-			args := []string{"mds", "deactivate", fmt.Sprintf("%s:%d", fsName, gid)}
-			if _, err := ExecuteCephCommand(context, clusterName, args); err != nil {
-				logger.Warningf(`Possibly failed to deactivate filesystem %s mds w/ gid %d
-In mimic+ this command does nothing; in luminous this is non-ideal but not necessarily critical: %v`,
-					fsName, gid, err)
-			}
+	// Deactivate any mdses with a higher rank than the desired max rank
+	// Ceph only allows mdses to be deactivated in reverse order starting with the highest rank
+	for gid := int(len(fs.MDSMap.In)) - 1; gid >= int(activeMDSCount); gid-- {
+		if err := deactivateMdsWithRetry(context, gid, clusterName, fsName); err != nil {
+			logger.Warningf("in mimic+ this error means nothing. in luminous this is non-ideal but not necessarily critical: %v", err)
+		}
+	}
+	// ** End of noted section 2
+
+	return nil
+}
+
+func deactivateMdsWithRetry(context *clusterd.Context, mdsGid int, namespace, fsName string) error {
+	retries := 10
+	retrySleep := 5 * time.Second
+	var err error
+	for i := 1; i <= retries; i++ {
+		args := []string{"mds", "deactivate", fmt.Sprintf("%s:%d", fsName, mdsGid)}
+		if _, err = ExecuteCephCommand(context, namespace, args); err == nil {
+			logger.Infof("successfully disabled mds with rank %d on attempt %d", mdsGid, i)
+			return nil
 		}
-	} // ** End of noted section 2
+		time.Sleep(retrySleep)
+	}
+	// report most recent error with additional err info
+	return fmt.Errorf("failed to deactivate mds w/ gid %d for filesystem %s: %+v", mdsGid, fsName, err)
+}
 
+// WaitForActiveRanks waits for the filesystem's number of active ranks to equal the desired count.
+// It times out with an error if the number of active ranks does not become desired in time.
+// Param 'moreIsOkay' will allow success condition if num of ranks is more than active count given.
+func WaitForActiveRanks(
+	context *clusterd.Context,
+	clusterName, fsName string,
+	desiredActiveRanks int32, moreIsOkay bool,
+	timeout time.Duration,
+) error {
+	ready := make(chan struct{})
+	quit := make(chan struct{})
+	defer close(ready)
+	defer close(quit)
+	go func() {
+		waitForActiveLoop(context, clusterName, fsName, int(desiredActiveRanks), moreIsOkay, ready, quit)
```

Comment:
I think this wait loop could be simplified by using a call to `wait.Poll()`. If you search for other uses in the rook repo, you'll see that it manages the retries and timeouts, and you don't need to do anything with goroutines or channels. Anytime we can avoid channels and goroutines, that's my recommendation. While they may seem simple once you get your head wrapped around them, they are still more complex and a source of bugs more than a simple polling loop.

Reply:
Definitely will do. I didn't know that existed, and I agree. Thanks.

Reply:
I was just reminded of it this morning when I was looking at something else, definitely was a good re-discovery.